drivers/misc/ethos_u/init.c

```diff
@@ -0,0 +1,159 @@
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(ethos_u, CONFIG_ARM_ETHOS_U_LOG_LEVEL);
+
+#define DT_DRV_COMPAT arm_ethos_u
+
+/*******************************************************************************
+ * Re-implementation/Overrides __((weak)) symbol functions from ethosu_driver.c
+ * To handle mutex and semaphores
+ *******************************************************************************/
+
+void *ethosu_mutex_create(void)
+{
+	struct k_mutex *mutex;
+
+	mutex = k_malloc(sizeof(*mutex));
+	if (mutex == NULL) {
+		LOG_ERR("Failed allocate mutex\n");
+		return NULL;
+	}
+
+	k_mutex_init(mutex);
+
+	return (void *)mutex;
+}
+
+int ethosu_mutex_lock(void *mutex)
+{
+	int status;
+
+	status = k_mutex_lock((struct k_mutex *)mutex, K_FOREVER);
+	if (status != 0) {
+		LOG_ERR("Failed to lock mutex with error - %d\n", status);
+		return -1;
+	}
+
+	return 0;
+}
+
+int ethosu_mutex_unlock(void *mutex)
+{
+	k_mutex_unlock((struct k_mutex *)mutex);
+	return 0;
+}
+
+void *ethosu_semaphore_create(void)
+{
+	struct k_sem *sem;
+
+	sem = k_malloc(sizeof(*sem));
+	if (sem == NULL) {
+		LOG_ERR("Failed to allocate semaphore\n");
+		return NULL;
+	}
+
+	k_sem_init(sem, 0, 100);
+
+	return (void *)sem;
+}
+
+int ethosu_semaphore_take(void *sem)
+{
+	int status;
+
+	status = k_sem_take((struct k_sem *)sem, K_FOREVER);
+	if (status != 0) {
+		LOG_ERR("Failed to take semaphore with error - %d\n", status);
+		return -1;
+	}
+
+	return 0;
+}
+
+int ethosu_semaphore_give(void *sem)
+{
+	k_sem_give((struct k_sem *)sem);
+	return 0;
+}
+
+struct ethosu_dts_info {
+	void *base_addr;
+	bool secure_enable;
+	bool privilege_enable;
+	void (*irq_config)(void);
+};
+
+struct ethosu_data {
+	struct ethosu_driver drv;
+};
+
+static int ethosu_zephyr_init(const struct device *dev)
+{
+	const struct ethosu_dts_info *config = dev->config;
+	struct ethosu_data *data = dev->data;
+	struct ethosu_driver *drv = &data->drv;
+	struct ethosu_driver_version version;
+
+	LOG_DBG("Ethos-U DTS info. base_address=0x%p, secure_enable=%u, privilege_enable=%u\n",
+		config->base_addr, config->secure_enable, config->privilege_enable);
+
+	ethosu_get_driver_version(&version);
+
+	LOG_DBG("Version. major=%u, minor=%u, patch=%u\n", version.major, version.minor,
+		version.patch);
+
+	if (ethosu_init(drv, config->base_addr, NULL, 0, config->secure_enable,
+			config->privilege_enable)) {
+		LOG_ERR("Failed to initialize NPU with ethosu_init().\n");
+		return -EINVAL;
+	}
+
+	config->irq_config();
+
+	return 0;
+}
+
+#define ETHOSU_DEVICE_INIT(n)									   \
+	static struct ethosu_data ethosu_data_##n;						   \
+												   \
+	static void ethosu_zephyr_irq_handler_##n(void)						   \
+	{											   \
+		struct ethosu_driver *drv = &ethosu_data_##n.drv;				   \
+		ethosu_irq_handler(drv);							   \
+	}											   \
+												   \
+	static void ethosu_zephyr_irq_config_##n(void)						   \
+	{											   \
+		IRQ_DIRECT_CONNECT(DT_INST_IRQN(n),						   \
```

Comment:
question here (not blocking): is there any reason to use DIRECT IRQs?

Reply:
Let's leave it like this for now.